tensorflow/java/src/test/java/org/tensorflow/ShapeTest.java

```diff
@@ -77,4 +80,36 @@ public void nodesInAGraph() {
       assertEquals(5, n.shape().size(1));
     }
   }
+
+  @Test
+  public void equalsWorksCorrectly() {
+    Shape s1 = Shape.make(-1, 2, 3);
+    Shape s2 = s1;
+    Shape s3 = Shape.make(-1, 2, 3);
+    Shape s4 = Shape.make(-1, 2, 4);
+    Shape s5 = Shape.make(-1, 2, 3, 4);
+    Shape s6 = Shape.make(-1, 2);
+    Object o = new Object();
+
+    assertEquals(s1, s2);
+    assertEquals(s1, s3);
+    assertNotEquals(s1, s4);
+    assertNotEquals(s1, s5);
+    assertNotEquals(s1, s6);
+    assertNotEquals(s1, o);
+    assertNotEquals(s1, null);
+  }
+
+  @Test
+  public void hashCodeIsAsExpected() {
+    long[] d1 = new long[] {1, 2, 3, 4};
+    long[] d2 = new long[] {};
+
+    Shape s1 = Shape.make(1, 2, 3, 4);
+    Shape s2 = Shape.scalar();
+
+    assertEquals(Arrays.hashCode(d1), s1.hashCode());
+    assertEquals(Arrays.hashCode(d2), s2.hashCode());
```

Comment:
These assertions are testing the implementation rather than behavior. Let's modify them to do the latter, so something like:

```java
assertEquals(Shape.make(1, 2, 3, 4).hashCode(), Shape.make(1, 2, 3, 4).hashCode());
assertNotEquals(Shape.make(1, 2).hashCode9), Shape.make(1, 3).hashCode());
```